Batched driver move abstraction

camelto2 opened the pull request:
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

This PR adds an abstraction layer to the batched drivers for the particle moves. It is intended to keep the VMC/DMC advanceWalkers agnostic about the actual moves that are being done (just particle positions, positions and spins, etc). Right now advanceWalkers is templated on a move type, and uses the templated class MoveAbstraction to handle the specifics of the moves. I am happy to change any names for classes/functions/etc here to make things more clear, I was more concerned with getting something working.

note: Eventually this should be able to handle the work needed for the L2potentials as well

part of #3712, and would allow me to delete #3742

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
macOSX

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

camelto2 commented:
Need to follow up and add comments to everything

camelto2 commented:
> Need to follow up and add comments to everything

I added documentation detailing the intention of each method in the MoveAbstraction class and describing each member variable. Also updated the docs to include the spin mass in the batched driver sections

ye-luo commented:
Here is my thoughts
1. MoveAbstraction constructor should capture ps_dispatcher and walker_elecs. Leave twf_dispatcher out.
2. Pass twf_dispatcher explicitly in APIs need it.
3. There should no trace of Crowd object in MoveAbstraction
4. acceptMove should be wrapped symmetrically to makeMove.

Replace mw_makeMoveWithSpin with mw_makeSpinMove which only deal with spin part. Call flex_makeMove and mw_makeSpinMove in the abstraction.
Add mw_accept_rejectSpinMove. Move the touch of spins from mw_accept_rejectMove to this new function and call it from the wrapper.

still thinking about a better name for MoveAbstraction. MoveManipulator? The object name should be a noun, not "move". Will can do the change once we settle the class name.

PDoakORNL reviewed (dismissed):
This PR sets up a new state machine and mixed data and operations class.
Due to the large upfront capture of variables and state members it hides the actual data dependence of many steps of the qmc section which I consider to be undesirable.
The methods are almost all state transformations of the object and not clean functions.

While I feel that breaking up advance walkers into pure functions would be worth while I think templating or giving them if constexpr sections based the Monte Carlo Coordinate type is a much cleaner way to add this functionality.

I would propose working from a different start such as in #3767

ye-luo commented:
@PDoakORNL can we first complete this one and then work on refactoring? #3767 is a prototype and only partially address what is abstracted in this PR. So it takes a bit more time to work out a fully satisfactory design.

camelto2 commented on src/QMCDrivers/MoveAbstraction.h:
```diff
@@ -0,0 +1,444 @@
+#ifndef QMCPLUSPLUS_MOVEABSTRACTION_H
+#define QMCPLUSPLUS_MOVEABSTRACTION_H
+
+#include "QMCDriverNew.h"
```
done

camelto2 commented:
> Here is my thoughts
> 
>     1. MoveAbstraction constructor should capture ps_dispatcher and walker_elecs. Leave twf_dispatcher out.
> 
>     2. Pass twf_dispatcher explicitly in APIs need it.
> 
>     3. There should no trace of Crowd object in MoveAbstraction

This is taken care of with the recent push. Looking at the changes to particleset now.

ye-luo commented on src/QMCDrivers/VMC/VMCBatched.cpp:
```diff
@@ -233,7 +213,13 @@ void VMCBatched::runVMCStep(int crowd_id,
   // For VMC we don't call this method for warmup steps.
   const bool accumulate_this_step = true;
-  advanceWalkers(sft, crowd, timers, *context_for_steps[crowd_id], recompute_this_step, accumulate_this_step);
+  const bool spin_move            = crowd.get_walker_elecs()[0].get().isSpinor();
```
This line can be dangerous if there is no element.
Pass the state via the sft object.
Also use `population_.get_golden_electrons()`

camelto2 replied:
done

ye-luo commented on src/QMCDrivers/VMC/VMCBatched.cpp:
```diff
@@ -80,68 +78,37 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
   std::vector<std::reference_wrapper<TrialWaveFunction>> twf_accept_list, twf_reject_list;
   isAccepted.reserve(num_walkers);
 
+  MoveAbstraction<COORDS> move(ps_dispatcher, walker_elecs, step_context.get_random_gen(), sft.drift_modifier,
```
Let us change this to 'mover' a noun.

camelto2 replied:
done

ye-luo reviewed (dismissed):
LGTM. I'm inclined to merge this for fast turnaround and avoid working on forks although I consider MoveAbstraction should go away as we get #3767 ready.

camelto2 commented:
Just incorporated the newer MCCoords stuff into this PR. @ye-luo also wants me also to template the ParticleSet mw_makeMove APIs on the CoordsType, so I will follow up with that with additional commits

camelto2 commented:
Made a significant change to this PR. This should be more in line with what @PDoakORNL was thinking.

I completely removed the MoveAbstraction class which this PR introduced.
Instead, I have introduced a new struct TWFGrads, similar to how MCCoords was introduced in #3774, which gets templated on the CoordsType. This allows me to carry around TWFGrads<CT> grads_now, grads_new in the advanceWalkers which will carry spatial gradients and spin gradients depending on the CoordsType. This also allowed me to template the TWFdispatcher on CoordsType, so that I can use a templated  flex_ API to decide if it is including spin or not for the gradients.
I also templated  DriftModifierBase which removes a lot of the divergent code for normal vs. spin calculations in advanceWalkers. 

With a little more work, the advanceWalkers will be further simplified so it doesn't have to have any "if constexpr (CT==CoordsType::POS_SPIN)" code

PDoakORNL commented:
Will start rereview this afternoon.

PDoakORNL commented on src/QMCDrivers/TauParams.hpp:
```diff
@@ -0,0 +1,65 @@
+//
+// Copyright (c) 2022 QMCPACK developers.
+//
+// File developed by: Ye Luo, yeluo@anl.gov, Argonne National Laboratory
```
@ye-luo didn't write this.
Put my name, yours and we have been writing
`// Refactored from MCCoords.hpp`
instead of File created by.

I think that is more meaningful when you're tearing up or porting a well known file.
The name is better.

ye-luo replied:
I did the refactoring. So I will take care of this.

PDoakORNL reviewed (approved):
This looks like a good step.  If there is time fix the attribution for TauParams.

ye-luo commented on src/QMCWaveFunctions/tests/test_MO_spinor.cpp:
```diff
@@ -271,8 +275,11 @@ void test_lcao_spinor()
     std::vector<ParticleSet::SingleParticlePos> displs = {dR[iat], dR[iat]};
+    std::vector<ParticleSet::Scalar_t> sdispls = {dS[iat], dS[iat]};
     elec_.mw_makeMove(p_list, iat, displs);
+    elec_.mw_makeSpinMove(p_list, iat, sdispls);
```
Could update this test to call the new APIs.

camelto2 replied:
yea, I was just working on this

camelto2 replied:
fixed

camelto2 commented:
ok, added a unit test for the TWFGrads class I added. With the current changes, there are only a few spots left in VMCBatched and DMCBatched that need to be abstracted, but those will be handled by a subsequent PR. This one is large enough as it is

ye-luo commented:
Test this please

ye-luo commented:
Test this please
